libsql_client/dbapi2/types.py

```diff
@@ -0,0 +1,1423 @@
+        itr = iter(iter_sql_statements(sql))
+        self.tokens = next(itr)
+        assert self.tokens
+        next_stmt = next(itr, None)
+        if next_stmt and next_stmt[0].upper() != "END":
+            msg = "You can only execute one statement at a time."
+            raise ProgrammingError(msg)
+        self.is_dml = self._check_is_dml(self.tokens)
+
+    @classmethod
+    def _check_is_dml(cls, tokens: Sequence[str]) -> bool:
+        # NOTE: this matches C pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L75-L82
+        # but it's not fully correct as the language allows: WITH ...
+        # See https://www.sqlite.org/lang_insert.html
+        # https://www.sqlite.org/lang_update.html
+        # https://www.sqlite.org/lang_delete.html
+        # Syntax:
+        #    WITH RECURSIVE table_name ( col_name, other_col)
+        #    AS NOT MATERIALIZED ( SELECT ... )
+        return tokens[0].upper() in cls._dml_statements
+
+    @property
+    def is_readonly(self) -> bool:
+        # mimics sqlite3_stmt_readonly()
+        return not self.is_dml
+
+
+def check_thread(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` thread usage.
+
+    Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_thread()
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+def check_connection(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` being valid.
+
+        Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_connection()
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+class Connection(metaclass=ABCMeta):
+    """Implement :py:class:`sqlite3.Connection` for remote servers.
+
+    :meta private:
+    """
+    _closed: bool
+    _thread: Optional[int]
+    _database: str
+    _timeout: float
+    _detect_types: int
+    _isolation_level: Optional[IsolationLevel]
+    _check_same_thread: bool
+    _cached_statement_getter: Callable[[str], Statement]
+    _in_transaction: bool
+    _cursors: WeakSet
+    _autocommit: Autocommit
+    _trace_callback: Optional[TraceCallback]
+    _limits: Dict[int, int]
+    row_factory: Optional[RowFactory]
+    text_factory: Optional[TextFactory]
+    cursor_factory: CursorFactory
+
+    def __init__(
+        self,
+        database: str,
+        timeout: float,
+        detect_types: int,
+        isolation_level: Optional[IsolationLevel],
+        check_same_thread: bool,
+        cached_statements: int,
+        autocommit: Autocommit,
```

Comment:
What is this `autocommit` parameter? I can't find it in the original sqlite3.

Reply:
yeah, not documented but exists in their code, is checked by their test suite and greatly impacts the behavior :-/